src/ninja.cc

```diff
@@ -514,6 +516,218 @@ int NinjaMain::ToolDeps(const Options* options, int argc, char** argv) {
   return 0;
 }
 
+class MissingDependencyScanner {
+ public:
+  MissingDependencyScanner(DepsLog* deps_log, State* state,
+                           DiskInterface* disk_interface,
+                           const string& gen_dir);
+  void ProcessNode(Node* node);
+  int PrintSummary();
+
+ private:
+  void ProcessNodeDeps(Node* node, Node** dep_nodes, int dep_nodes_count);
+  bool IsNodeInGenDir(Node* node);
+
+  DepsLog* deps_log_;
+  State* state_;
+  DiskInterface* disk_interface_;
+  set<Node*> seen_;
+  set<Node*> nodes_missing_deps_;
+  set<Node*> generated_nodes_;
+  set<const Rule*> generator_rules_;
+  int missing_dep_path_count_;
+  set<Node*> nodes_missing_gen_;
+  string abs_gen_dir_;
+};
+
+MissingDependencyScanner::MissingDependencyScanner(
+    DepsLog* deps_log, State* state, DiskInterface* disk_interface,
+    const string& gen_dir)
+    : deps_log_(deps_log), state_(state), disk_interface_(disk_interface),
+      missing_dep_path_count_(0) {
+  if (!gen_dir.empty()) {
+    char buf[PATH_MAX];
+    if (getcwd(buf, PATH_MAX) == NULL)
+      Fatal("getcwd failed");
+    abs_gen_dir_ = buf;
+    abs_gen_dir_.append("/");
+    abs_gen_dir_.append(gen_dir);
+    if (!realpath(abs_gen_dir_.c_str(), buf))
```

Comment:
msvc/mingw don't like the realpath calls (AppVeyor build failed). Any recommendations?

Reply:
You can use `std::filesystem::canonical`: https://en.cppreference.com/w/cpp/filesystem/canonical

You can use `#if _MSC_VER > 1910` to detect if its available and otherwise disable your tool (not really a problem, as our Windows release is build with Visual Studio 2017).

Reply:
Hm, not that I mind using C++17, but conditionally disabling the tool may be fairly ugly (no tools do that as far as I see?)

I noticed there's a Windows-specific AbsPath in src/includes_normalize.h, I can try moving it to src/util.h/cc so I can use it in the tool (may have to move the currently-local  InternalGetFullPathName as well to avoid changing too much other code), I think that'd be preferable over #including a weird header in a dependency analysis tool.

Reply:
The browse tool is disabled on Windows and on any platform when `NINJA_PYTHON` isn't available at runtime.

Reply:
oh, thanks. I'll try it out and see how it goes, time permitting.

Reply:
this is now obsolete because the -g tool that required abspath is no longer part of this pull request to keep it more manageable